imapclient/client.go

```diff
@@ -0,0 +1,768 @@
+	br       *bufio.Reader
+	bw       *bufio.Writer
+	dec      *imapwire.Decoder
+	enc      *imapwire.Encoder
+	encMutex sync.Mutex
+
+	mutex       sync.Mutex
+	cmdTag      uint64
+	pendingCmds []command
+	contReqs    []chan struct{}
+}
+
+// New creates a new IMAP client.
+//
+// This function doesn't perform I/O.
+func New(conn net.Conn) *Client {
+	var (
+		r io.Reader = conn
+		w io.Writer = conn
+	)
+	if debug {
+		r = io.TeeReader(r, os.Stderr)
+		w = io.MultiWriter(w, os.Stderr)
+	}
+
+	br := bufio.NewReader(r)
+	bw := bufio.NewWriter(w)
+
+	client := &Client{
+		conn: conn,
+		br:   br,
+		bw:   bw,
+		dec:  imapwire.NewDecoder(br),
+		enc:  imapwire.NewEncoder(bw),
+	}
+	go client.read()
+	return client
+}
+
+// DialTLS connects to an IMAP server with implicit TLS.
+func DialTLS(address string) (*Client, error) {
+	conn, err := tls.Dial("tcp", address, nil)
+	if err != nil {
+		return nil, err
+	}
+	return New(conn), nil
+}
+
+// Close immediately closes the connection.
+func (c *Client) Close() error {
+	return c.conn.Close()
+}
+
+// beginCommand starts sending a command to the server.
+//
+// The command name and a space are written.
+//
+// The caller must call endCommand.
+func (c *Client) beginCommand(name string) *Command {
+	c.encMutex.Lock() // unlocked by endCommand
+
+	c.mutex.Lock()
+	c.cmdTag++
+	tag := fmt.Sprintf("T%v", c.cmdTag)
+	c.mutex.Unlock()
+
+	cmd := &Command{
+		tag:  tag,
+		done: make(chan error, 1),
+		enc:  c.enc,
+	}
+	cmd.enc.Atom(tag).SP().Atom(name)
+	return cmd
+}
+
+// endCommand ends an outgoing command.
+//
+// A CRLF is written.
+//
+// The command is registered as a pending command until the server sends a
+// completion result response.
+func (c *Client) endCommand(cmd command) {
+	baseCmd := cmd.base()
+
+	c.mutex.Lock()
+	c.pendingCmds = append(c.pendingCmds, cmd)
+	c.mutex.Unlock()
+
+	if err := baseCmd.enc.CRLF(); err != nil {
+		baseCmd.err = err
+	}
+	baseCmd.enc = nil
+	c.encMutex.Unlock()
+}
+
+func (c *Client) deletePendingCmdByTag(tag string) command {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for i, cmd := range c.pendingCmds {
+		if cmd.base().tag == tag {
+			c.pendingCmds = append(c.pendingCmds[:i], c.pendingCmds[i+1:]...)
+			return cmd
+		}
+	}
+	return nil
+}
+
+func findPendingCmdByType[T interface{}](c *Client) T {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	for _, cmd := range c.pendingCmds {
+		if cmd, ok := cmd.(T); ok {
+			return cmd
+		}
+	}
+
+	var cmd T
+	return cmd
+}
+
+func (c *Client) registerContReq() <-chan struct{} {
+	ch := make(chan struct{})
+
+	c.mutex.Lock()
+	c.contReqs = append(c.contReqs, ch)
+	c.mutex.Unlock()
+
+	return ch
+}
+
+func (c *Client) unregisterContReq(ch <-chan struct{}) {
+	c.mutex.Lock()
+	for i := range c.contReqs {
+		if (<-chan struct{})(c.contReqs[i]) == ch {
+			c.contReqs = append(c.contReqs[:i], c.contReqs[i+1:]...)
+			break
+		}
+	}
+	c.mutex.Unlock()
+}
+
+func (c *Client) encodeLiteral(size int64) io.WriteCloser {
```

Comment:
If the server doesn't reply with a continuation request, but instead replies with an error, we need to unregister the continuation request.